### PR TITLE
Rabbit Holes (feature branch)

### PR DESCRIPTION
**Pull request by dbrant**
Implementation of the Rabbit Holes experiment.

**Phabricator:**
https://phabricator.wikimedia.org/T379379
https://phabricator.wikimedia.org/T379381
https://phabricator.wikimedia.org/T379382

### CODE REVIEW

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt [hunk 1]
```diff
@@ -62,12 +67,19 @@ class RecentSearchesFragment : Fragment() {
...
+        binding.recentSearchesRecycler.adapter = if (suggestedSearchTerm.isNullOrEmpty()) RecentSearchAdapter() else ConcatAdapter(SuggestedSearchAdapter(), RecentSearchAdapter())
+        binding.recentSearchesHeaderContainer.isVisible = suggestedSearchTerm.isNullOrEmpty()
+
         val touchCallback = SwipeableItemTouchHelperCallback(requireContext())
         touchCallback.swipeableEnabled = true
         val itemTouchHelper = ItemTouchHelper(touchCallback)
         itemTouchHelper.attachToRecyclerView(binding.recentSearchesRecycler)
         setButtonTooltip(binding.recentSearchesDeleteButton)
+
+        RabbitHolesEvent.submit("impression", "search")
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt:81]: Do we want to check whether the rabbit hole is enabled?

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt [hunk 2]
```diff
@@ -123,30 +134,38 @@ class RecentSearchesFragment : Fragment() {
...
 
-        fun bindItem(position: Int) {
+        open fun bindItem(position: Int) {
             recentSearch = recentSearchList[position]
             itemView.setOnClickListener(this)
             (itemView as TextView).text = recentSearch.text
         }
 
         override fun onClick(v: View) {
+            RabbitHolesEvent.submit("recent_search_click", "search",
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt:166]: Do we want to check whether the rabbit hole is enabled?

File: app/src/main/java/org/wikipedia/search/SearchFragment.kt [hunk 3]
```diff
@@ -205,7 +209,11 @@ class SearchFragment : Fragment(), SearchResultsFragment.Callback, RecentSearche
             requireActivity().setResult(SearchActivity.RESULT_LINK_SUCCESS, intent)
             requireActivity().finish()
         } else {
-            val historyEntry = HistoryEntry(item, HistoryEntry.SOURCE_SEARCH)
+
+            RabbitHolesEvent.submit("navigate", "search",
+                source = if (query == suggestedSearchQuery) "suggested" else "default")
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/search/SearchFragment.kt:214]: Do we want to check whether the rabbit hole is enabled?

**Review by cooltey — CHANGES_REQUESTED** [review]
Did a quick review and will do another round of review today (before you off from work)

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt (hunk 1 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt:81]: Whoops, I see what you mean; I've updated the RabbitHolesEvent send itself only if the experiment is active.

**Review by Williamrai — APPROVED** [review]

**Review by cooltey — CHANGES_REQUESTED** [review]
Found issues:
- When on the suggested reading list screen, it is possible to swipe to refresh. Should we disable it? 
- The "experiment" card overlapped when on the search mode in the reading list. (happens when no results found in the list)
- When trying to Save the recommended list, the articles that don't have images will still show the placeholder. Do we want to hide it instead?
- For analytics, I am not able to monitor the `app_rabbit_holes` from https://stream-beta.wmflabs.org/v2/ui/#/ is this expected?

File: app/src/main/java/org/wikipedia/page/PageActivity.kt [hunk 4]
```diff
@@ -794,6 +820,110 @@ class PageActivity : BaseActivity(), PageFragment.Callback, LinkPreviewDialog.Lo
...
+    private fun maybeShowRabbitHolesSurvey() {
+        if (Prefs.suggestedContentSurveyShown) {
+            return
+        }
+        lifecycleScope.launch(CoroutineExceptionHandler { _, t ->
+            L.e(t)
+        }) {
+            delay(TimeUnit.SECONDS.toMillis(if (ReleaseUtil.isDevRelease) 1L else 10L))
+            pageFragment.historyEntry?.let {
+                if (!Prefs.suggestedContentSurveyShown && it.source == HistoryEntry.SOURCE_RABBIT_HOLE_SEARCH) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/page/PageActivity.kt:913]: Can we move this to the line `905`? This seems duplicated.

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt [hunk 5]
```diff
@@ -160,6 +179,57 @@ class RecentSearchesFragment : Fragment() {
...
+        override fun getItemViewType(position: Int): Int {
+            return if (position == 1) LIST_TYPE_ITEM else LIST_TYPE_HEADING
+        }
+
+        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+            return if (viewType == LIST_TYPE_HEADING) {
+                SuggestedSearchHeadingViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false))
+            } else {
+                SuggestedSearchItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false)
+                )
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt:220]: These can be simplified:
```
val view = LayoutInflater.from(parent.context).inflate(R.layout.item_search_recent, parent, false)
return if (viewType == LIST_TYPE_HEADING) {
SuggestedSearchHeadingViewHolder(view)
} else {
SuggestedSearchItemViewHolder(view)
}
```

File: app/src/main/java/org/wikipedia/page/PageActivity.kt (hunk 4 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/page/PageActivity.kt:910]: Sure.

File: app/src/main/java/org/wikipedia/readinglist/ReadingListFragment.kt [hunk 6]
```diff
@@ -315,25 +328,42 @@ class ReadingListFragment : Fragment(), MenuProvider, ReadingListItemActionsDial
...
+                        }
+                    } else {
+                        val json = Prefs.receiveReadingListsData
+                        if (!json.isNullOrEmpty()) {
+                            readingList = ReadingListsReceiveHelper.receiveReadingLists(requireContext(), json, encoded = true)
+                            readingList?.let {
+                                ReadingListsAnalyticsHelper.logReceivePreview(requireContext(), it)
+                                binding.searchEmptyView.setEmptyText(getString(R.string.search_reading_list_no_results, it.title))
+                            }
+                            update()
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/readinglist/ReadingListFragment.kt]: Sure!

File: app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt (hunk 5 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/search/RecentSearchesFragment.kt]: Ok!

**Comment by dbrant** [comment]
@cooltey 
> * When on the suggested reading list screen, it is possible to swipe to refresh. Should we disable it?

Sure!

> * The "experiment" card overlapped when on the search mode in the reading list. (happens when no results found in the list)

Done. (The suggested reading list will be hidden from searching)

> * When trying to Save the recommended list, the articles that don't have images will still show the placeholder. Do we want to hide it instead?

This does not appear to be an issue introduced in this PR.

> * For analytics, I am not able to monitor the `app_rabbit_holes` from https://stream-beta.wmflabs.org/v2/ui/#/ is this expected?

This stream/schema has not yet been deployed. (This is not going to production yet.)

**Review by cooltey — APPROVED** [review]

**Pull request merged**